File: WdtcCore/src/main/java/org/wdt/wdtc/game/DetermineVersionSize.java
package org.wdt.wdtc.game;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class DetermineVersionSize {

    public static boolean DetermineSize(String OriginalVersionNumber, Launcher launcher) {
        Matcher OriginalVersion = getVersionNumber(OriginalVersionNumber);
        Matcher Version = getVersionNumber(launcher.getVersion());
        if (OriginalVersion.find() && Version.find()) {
            int Original = Integer.parseInt(OriginalVersion.group(1));
            int target = Integer.parseInt(Version.group(1));
            if (Original == target) {
                if (OriginalVersion.group(2) != null && Version.group(2) != null) {
                    int OriginalMinorVersionNumber = Integer.parseInt(OriginalVersion.group(2));
                    int MinorVersionNumber = Integer.parseInt(Version.group(2));
                    if (OriginalMinorVersionNumber == MinorVersionNumber) return true;
                    return OriginalMinorVersionNumber < MinorVersionNumber;
                } else if (OriginalVersion.group(2) != null) {
                    return false;
                } else if (Version.group(2) != null) {
                    return true;
                } else {
                    return true;
                }
            }
            return Original < target;

        }
        return false;
    }

    private static Matcher getVersionNumber(String str) {
        return Pattern.compile("1\\.([0-9][0-9])\\.?([0-9])?").matcher(str);
    }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/game/DetermineVersionSize.kt
package org.wdt.wdtc.core.game

import java.util.regex.Matcher
import java.util.regex.Pattern

// TODO Using a method of comparing release times
fun determineSize(originalVersionNumber: String, launcher: Launcher): Boolean {
  val originalVersion = getVersionNumber(originalVersionNumber)
  val version = getVersionNumber(launcher.versionNumber)
  if (originalVersion.find() && version.find()) {
    val original = originalVersion.group(1).toInt()
    val target = version.group(1).toInt()
    return if (original == target) {
      if (originalVersion.group(2) != null && version.group(2) != null) {
        val originalMinorVersionNumber = originalVersion.group(2).toInt()
        val minorVersionNumber = version.group(2).toInt()
        if (originalMinorVersionNumber == minorVersionNumber) true else originalMinorVersionNumber < minorVersionNumber
      } else if (originalVersion.group(2) != null) {
        false
      } else if (version.group(2) != null) {
        true
      } else {
        true
      }
    } else original < target
  }
  return false
}

private fun getVersionNumber(str: String): Matcher {
  return Pattern.compile("1\\.([0-9][0-9])\\.?([0-9])?").matcher(str)
}

